tokio/src/runtime/task/list.rs

```diff
@@ -179,22 +184,46 @@ impl<S: 'static> OwnedTasks<S> {
 
         // safety: We just checked that the provided task is not in some other
         // linked list.
-        unsafe { self.inner.lock().list.remove(task.header_ptr()) }
+        unsafe { self.remove_inner(task) }
+    }
+
+    #[inline]
+    unsafe fn remove_inner(&self, task: &Task<S>) -> Option<Task<S>> {
+        let mut lock = self.segment_inner(task.task_id() as usize);
+        match lock.remove(task.header_ptr()) {
+            Some(task) => {
+                drop(lock);
+                self.count.fetch_sub(1, Ordering::Relaxed);
+                Some(task)
+            }
+            None => None,
+        }
```

Comment:
style nit, take it or leave it: i think we can simplify this code to just
```suggestion
        let task = self.segment_inner(task.task_id() as usize).remove(task.header_ptr())?;
        self.count.fetch_sub(1, Ordering::Relaxed);
        Some(task)
```

Reply:
Yes! We use `?` here to resolve `None` better, but I would like to drop mutex manully here.